isic.1
.\" Process this file with
.\" groff -man -Tascii foo.1
.\" 
.\" MANEDIT
.TH "ISIC" "1" "JANUARY 2007" "v0.07" "User Manuals"
.SH "NAME"
ISIC - IP Stack Integrity Checker
.SH "SYNOPSIS"
.B esic -i
.I <interface>
.B [...]
.br 
.B isic -s 
.I <source ip> 
.B -d 
.I <destination ip> 
.B [...]
.br 
.B tcpsic -s 
.I <source ip>[,port]
.B -d 
.I <destination ip>[,port]
.B [...]
.br 
.B udpsic -s 
.I <source ip>[,port]
.B -d 
.I <destination ip>[,port]
.B [...]
.br 
.B icmpsic -s 
.I <source ip>
.B -d 
.I <destination ip>
.B [...]
.br 
.B multisic -s
.I <source ip>[,port]
.B -d
.I <destination ip>[,port]
.B -i
.I <interface>
.B [...]
.br
.B isic6 -s 
.I <source ip>
.B -d 
.I <destination ip>
.B [...]
.br 
.B tcpsic6 -s 
.I <source ip>[,port]
.B -d 
.I <destination ip>[,port]
.B [...]
.br 
.B udpsic6 -s 
.I <source ip>[,port]
.B -d 
.I <destination ip>[,port]
.B [...]
.br 
.B imcpsic6 -s 
.I <source ip>
.B -d 
.I <destination ip>
.B [...]
.br 
.SH "DESCRIPTION"
.B ISIC
is a suite of utilities to exercise the stability of an IP Stack 
and its component stacks (TCP, UDP, ICMP et. al.) It generates piles 
of pseudo random packets of the target protocol. The packets are given 
tendancies to conform to. For example 50% of the packets generated can
have IP Options, and 25% of the packets can be IP fragments... But the
percentages are arbitrary and most of the packet fields have a configurable
tendancy.
.PP 
The packets are then sent against the target machine to either penetrate
its firewall rules or find bugs in the IP stack. ISIC also contains an 
utility (esic) that generates raw Ethernet frames to examine hardware 
implementations, and the "multisic" tool to check multicast code integrity.
.PP 
Starting from version
.IR 0.07 ,
ISIC includes utilities (*sic6) to test IPv6 protocol
stack. These would of course need IPv6 enabled in the system as a prerequisite.
.SH "GENERAL OPTIONS"
.IP "-c <# of packets>"
For esic only, number of packets to send. It is equivlent to \-p option 
used by other utilities. The default is
.IR 2^32 .
.IP "-d <destination MAC/IP>"
Optional for esic, required for all others. For esic, this specifies 
destination MAC address, and the default will be broadcast address
.IR ff:ff:ff:ff:ff:ff .
For all others, this specifies destination IP address. Use "\-d rand" to send packets
to random destination MAC/IP address. For multisic, "\-d rand" will make destination
IP address picked from the list [224, 225, 232, 233, 234, 235, 236, 237, 238, 239]
randomly.
.IP "-i <interface>"
Required by esic and multisic. This specifies the interface for outgoing packets.
.IP "-k <num>"
This specifies the number of skipped packets. This option
is useful for identifying the exact packet(s) causing the problem. Default is 
.IR 0 .
.IP "-l <length>"
For esic only, the maximum length of packet generated. Default is maximum
ethernet frame size
.IR 1500 .
Length greater than default value will be rejected.
.IP "-m <num>"
For esic, this is the number of packets between printout, default to
.IR 1000 .
For all others, this specifies the maximum speed of traffic generation, unit is kB/s
.IP "-p <num>"
For esic, this specifies upper layer protocol number in the ethernet header, and
default to type IP 
.IR 0x0800 .
For all others this is the number of packets to send, default is
.IR 2^32 .
Note esic uses \-c instead for this purpose.
.IP "-r <random seed>"
Specify the random seed, default to current process id.
.IP "-s <source MAC/IP>"
Optional for esic, required for all others. For esic, this specifies
source MAC address, and the default will be the MAC address of given interface.
For all others, this specifies source IP address. Use "-s rand" to set source
MAC/IP address to random value.
.IP "-x <num>"
Available to all but esic. Send each packet <num> times, the default value is
.IR 1 .
.IP -v
Print out the current version of ISIC test suite.
.IP "-z <source MAC>"
Optional for multisic. This specifies source MAC address for outgoing multicast
packets, and the default will be the MAC address of given interface. Use "-z rand"
to set source MAC address to random value.
.IP -D
Available to all but esic, this turns on the debugging printout. Caution: 
the program will print out information of every generated packet, the entire
output will be huge.

.SH "PERCENTAGE OPTIONS"
These options are defined to all but esic. All percentage options have
default value of
.IR 10 .
.IP -i
Available for icmpsic and icmpsic6 only, the percentage of ICMP packets with
bad ICMP checksum.
.IP -t
Available for tcpsic and tcpsic6 only, the percentage of TCP packets with
bad TCP checksum.
.IP -u
Available for tcpsic and tcpsic6 only, the percentage of TCP packets with
urgent flag set.
.IP -F
The percentage of fragmented packets. Note for *sic6, this is the percentage
of IPv6 packets with random fragment extension header.
.IP -I
 For *sic6, this is the percentage of IPv6 packets with random destination
option header. For isic, this specifies the percentage of packets with
random IP header length. For tcpsic, udpsic and icmpsic, this is the
percentage of packets with random IP option.
.IP -T
Available for tcpsic and tcpsic6 only, the percentage of TCP options in
the TCP header.
.IP -U
Available for udpsic and udpsic6 only, the percentage of UDP packets with
bad UDP checksum.
.IP -V
Avialable for isic, tcpsic, udpsic, icmpsic and isic6, this is the percentage
of packets with bad IP version value in the header.

.SH "EXAMPLES"
.IP "esic -i eth0 -s 01:02:34:56:07:89 -p rand -m 5000"
.IP
.br
.RS
This will generate ethernet frames with random protocol number in the ethernet
header, and send out through eth0 interface. In the frames, the source
MAC address is fixed 01:02:34:56:07:89, destination will be the default broadcast
MAC address. There will be a printout line for every 5000 frames.
.RE
.IP "isic -s rand -d 10.11.12.13 -F 50 -p 100 -k 20 -r 10"
.IP
.RS
This asks isic to generate 100 IP packets with randsom source address
and fixed destination address 10.11.12.13. The random seed is set to 10.
half of packets will be fragments. When sending out, first 20 packets will
be skipped, isic will start from the 21st packet.
.RE
.IP "tcpsic -s 1.2.3.4,69 -d 21.22.23.24 -x 2 -m 1000 -T 30 -t 50"
.IP
.RS
This advises tcpsic to generate TCP packets with source address 1.2.3,4 and
source TCP port 69, and destination address 21.22.23.24 and random destination
TCP port. Each packet will be sent out twice, and the overall maximum speed is
1000kB/s. Of all the TCP packets generated, 30% of packets will have random
TCP options, and 50% will have bad TCP checksum.
.RE
.IP "multisic -s rand -d 224.0.0.5 -i eth2 -p 50000 -F 50 -z ff:ff:ff:ff:ff:ff"
.IP
.RS
This lets multisic to send 50000 UDP packets to multicast address 224.0.0.5 with 
random source address and source/destination UDP ports. The egress interface is 
forced to be eth2. 50% of outgoing packets will have fragments. And the source
MAC address is set to ff:ff:ff:ff:ff:ff.
.RE
.IP "udpsic6 -s rand -d  2001:1:2:3:4::2,161 -p 1000000 -I 90 -U 20"
.IP
.RS
This lets udpsic6 to send 1 million IPv6 UDP packets with random source address
and source UDP port, to destination address 2001:1:2:3:4::2 and UDP port 161 (SNMP port).
90% of outgoing packets will have random IPv6 destination option header, and 20% of
total packets will include incorrect UDP checksum.
.RE


.SH "AUTHOR"
Shu Xiao <sxiao@cisco.com>
.SH "SEE ALSO"
Libnet (http://www.packetfactory.net/projects/libnet)
.br 
ISIC online (http://www.packetfactory.net/projects/ISIC) or
            (http://isic.sourceforge.net)
.br 
ip6sic (http://ip6sic.sourceforge.net)


ChangeLog

ISIC (v0.07)	12/22/2006, by Shu Xiao (sxiao@cisco.com)

	- New IPv6 utilities: isic6, tcpsic6, udpsic6, icmpsic6.
	
	- New tool multisic for sending random multicast UDP packets. 
	
	- Fixed compilation errors with GCC v4.

	- Added -k skip option for esic per the request from some users.

	- Placed the flood control under low traffic limit case only.

		The original flood protection code calls gettimeofday(),
		and runs after sending each packet. This is time consumping
		and prevents the tool from reaching high traffic limit
		on high speed interface (eg. 1000Base-TX). Now the flood
		control will take effect only if the specified maximum rate
		is less than 10MB/s.

	- Corrected -t option bug with tcpsic:
	
		The old one was using it reversely, e.g. -t 30 would generate
		packet with 30% of them having correct TCP checksum. But -t 
		is actually for the percentage of bad TCP checksum. 

	- Changed to used 32-bit random numbers for some fields:
	
		For IPv4 *sic tools, the source/destination IP address and
		TCP sequence/acknowledge numbers are 32-bit width, and
		should be assigned with equal size random numbers. The old
		code did not do shifting correctly.

	- Changed the default value to 10 for all percentage options.
 
	- Cleaned the usage printout of icmpsic:

		There were ports used in the usage printout of icmpsic. This
		was misleading since no ports arguments needed for it.

	- New manual page, try to get more help with "man isic".


ISIC (v0.06)	11/10/2004, by Shu Xiao (sxiao@cisco.com)

	- Changed to link with Libnet 1.1.x:

		With minimum code change to replace old calls to Libnet 1.0.x
		with Libnet 1.1.x functions, and keep Mike's major fuzzing 
		algorithm untouched.
		
		Also found that v0.05 had checksum error with esic, which 
		rooted from bug of Libnet 1.0.x. v0.06 esic works fine with 
		Libnet 1.1.x 

	- More accurate datapush rate calculation: 

		In v0.05 isic, tcpsic, udpsic and icmpsic code, the return
		value of the call to packet sending function was not checked.
		But it actually could fail and return -1, in which case the
		datapush byte count was calculated wrong. v0.06 code skips
		failed cases of data pushing.
 	
	- Fixed esic MAC address printout bug:

		v0.05 esic has a minor problem with MAC address printout.
		The source and destination MAC addresses was actually reversed.



isic.c
#include "isic.h"

/*
 * ISIC - IP Stack Integrity Checker
 *
 * This is tuned for ethernet sized frames (1500 bytes)
 * For user over a modem or frame (or other) you will have to change the
 * 'rand() & 0x4ff' line below.  The 0x4ff needs to be less than the size of
 * the frame size minus the length of the ip header (20 bytes IIRC)
 */


/* Variables shared between main and the signal handler so we can display
 * output if ctrl-c'd
 */
u_int seed = 0;
u_long acx = 0;
struct timeval starttime;
u_long datapushed = 0;			/* How many bytes we pushed */


int
main(int argc, char **argv)
{
	int c;
	u_char *buf = NULL;
	u_short	*payload = NULL;
	u_int payload_s = 0;

	/* libnet variables */
	char errbuf[LIBNET_ERRBUF_SIZE];
	libnet_t *l;
	char *device = NULL;

	/* Packet Variables */
	struct ip *ip_hdr = NULL;
	u_int32_t src_ip = 0, dst_ip = 0;
	u_char tos, ttl, ver;
	u_int id, prot, frag_off, hl;


	/* Functionality Variables */
	int src_ip_rand = 0, dst_ip_rand = 0;
	struct timeval tv, tv2;
	float sec;
	unsigned int cx = 0;
	u_long max_pushed = 10240;		/* 10MB/sec */
	u_long num_to_send = 0xffffffff;	/* Send 4billion packets */
	u_long skip = 0; 			/* Skip how many packets */
	int printout = 0;			/* Debugging */
	u_int repeat = 1;			/* How many times to send
						 *  each packet */

	/* Defaults */
	float FragPct	=	10;
	float BadIPVer	=	10;
	float IPLength	=	10;

	/* Not crypto strong randomness but we don't really care.  And this  *
	 * gives us a way to determine the seed while the program is running *
 	 * if we need to repeat the results				     */
	seed = getpid();

	/* Initialize libnet context, Root priviledges are required.*/ 
	l = libnet_init(
            LIBNET_RAW4_ADV,                        /* injection type */
            device,                                 /* network interface */
            errbuf);                                /* error buffer */

	if (l == NULL) {
	  fprintf(stderr, "libnet_init() failed: %s", errbuf);
	  exit( -1 );
	}
                
	while((c = getopt(argc, argv, "hd:I:s:r:m:k:Dp:V:F:vx:")) != EOF) {
	  switch (c) {
	   case 'h':
		usage(argv[0]);
		exit(0);
		break;
	   case 'd':
		if ( strncmp(optarg, "rand", sizeof("rand")) == 0 ) {
			printf("Using random dest IP's\n");
			dst_ip = 1;	/* Just to pass sanity checks */
			dst_ip_rand = 1;
			break;
		}
		if ((dst_ip = libnet_name2addr4(l, optarg, LIBNET_RESOLVE)) == (u_int32_t)-1) {
			fprintf(stderr, "Bad dest IP\n");
			exit( -1 );
		}
		
		break;
	   case 's':
		if ( strncmp(optarg, "rand", sizeof("rand")) == 0 ) {
			printf("Using random source IP's\n");
			src_ip = 1;	/* Just to pass sanity checks */
			src_ip_rand = 1;
			break;
		}
		if ((src_ip = libnet_name2addr4(l, optarg, LIBNET_RESOLVE)) == (u_int32_t)-1) {
			fprintf(stderr, "Bad source IP\n");
			exit( -1 );
		}
		break;
	   case 'r':
		seed = atoi(optarg);
		break;
	   case 'm':
		max_pushed = atol(optarg);
		break;
	   case 'k':
		skip = atol(optarg);
		printf("Will not transmit first %li packets.\n", skip);
		break;
	   case 'D':
		printout++;
		break;
	   case 'p':
		num_to_send = atoi(optarg);
		break;
	   case 'V':
		BadIPVer = atoi(optarg);
		break;
	   case 'F':
		FragPct = atof(optarg);
		break;
	   case 'I':
		IPLength = atof(optarg);
		break;
	   case 'x':
		repeat = atoi(optarg);
		break;
	   case 'v':
		printf("Version %s\n", VERSION);
		exit(0);
	   }
	}

	if ( !src_ip || !dst_ip ) {
		usage(argv[0]);
		exit(EXIT_FAILURE);
	}

	printf("Compiled against Libnet %s\n", LIBNET_VERSION);
	printf("Installing Signal Handlers.\n");
	if ( signal(SIGTERM, &sighandler) == SIG_ERR )
		printf("Failed to install signal handler for SIGTERM\n");
	if ( signal(SIGINT, &sighandler) == SIG_ERR )
		printf("Failed to install signal handler for SIGINT\n");
	if ( signal(SIGQUIT, &sighandler) == SIG_ERR )
		printf("Failed to install signal handler for SIGQUIT\n");

	printf("Seeding with %i\n", seed);
	srand(seed);
	max_pushed *= 1024;

	if ( (buf = malloc(IP_MAXPACKET)) == NULL ) {
	  perror("malloc: ");
	  exit( -1 );
	}

        if ( max_pushed >= 10000000 )
	  printf("No Maximum traffic limiter\n");
        else printf("Maximum traffic rate = %.2f k/s\n", max_pushed/1024.0 );

        printf("Bad IP Version\t= %.0f%%\t\t", BadIPVer);
        printf("Odd IP Header Length\t= %.0f%%\t\t", IPLength);
        printf("Frag'd Pcnt\t= %.0f%%\n", FragPct);


	/* Drop them down to floats so we can multiply and not overflow */
	BadIPVer	/= 100;
	FragPct		/= 100;
	IPLength	/= 100;

    

	/*************
 	* Main Loop *
 	*************/
	gettimeofday(&tv, NULL);
	gettimeofday(&starttime, NULL);

	for(acx = 0; acx < num_to_send; acx++) {
	
		tos	= RAND8;
		id	= acx & 0xffff;
		ttl	= RAND8;
		prot	= RAND16;

		if ( rand() <= (RAND_MAX * FragPct) )
			frag_off = RAND16;
		else	frag_off = 0;

		if ( src_ip_rand == 1 )
		  src_ip = RAND32;
		if ( dst_ip_rand == 1 )
		  dst_ip = RAND32;

		if ( rand() <= (RAND_MAX * BadIPVer ) )
			ver = rand() & 0xf;
		else	ver = 4;

		if ( rand() <= (RAND_MAX * IPLength ) )
			hl = rand() & 0xf;
		else	hl = 5;

		payload_s = rand() & 0x4ff;            /* length of 1279 */
		
		/* Build the IP header */
		ip_hdr = (struct ip *) buf;
		ip_hdr->ip_v    = ver;                 /* version 4 */
		ip_hdr->ip_hl   = hl;                  /* 20 byte header */
		ip_hdr->ip_tos  = tos;                 /* IP tos */
		ip_hdr->ip_len  = htons(IP_H + payload_s);  /* total length */
		ip_hdr->ip_id   = htons(id);           /* IP ID */
		ip_hdr->ip_off  = htons(frag_off);     /* fragmentation flags */
		ip_hdr->ip_ttl  = ttl;                 /* time to live */
		ip_hdr->ip_p    = prot;                /* transport protocol */
		ip_hdr->ip_sum  = 0;                   /* do this later */
		ip_hdr->ip_src.s_addr = src_ip;
		ip_hdr->ip_dst.s_addr = dst_ip;
		
		payload = (u_short *)(buf + IP_H);
		for(cx = 0; cx <= (payload_s >> 1); cx+=1)
			payload[cx] = RAND16;
		payload[payload_s] = RAND16;
		
		if ( printout ) {
			printf("%s ->",
				inet_ntoa(*((struct in_addr*) &src_ip )));
			printf(" %s tos[%i] id[%i] ver[%i] frag[%i]\n",
				inet_ntoa(*((struct in_addr*) &dst_ip )), tos,
				id, ver, frag_off);
		}
			
		if ( skip <= acx ) {
		  for ( cx = 0; cx < repeat; cx++ ) {
		    c = libnet_write_raw_ipv4(l, buf, IP_H + payload_s);
		    if (c != -1)
		      datapushed+=c;
		  }
	/*  This is way too noisy!
	 *		if (c < (signed) (IP_H + payload_s)) {
	 *			perror("sendto: ");
	 *			fprintf(stderr, "Packet Failed '%i'\n", c);
	 *			printf("%s ->", 
	 *			     inet_ntoa(*((struct in_addr*) &src_ip )));
	 *			printf(" %s tos[%i] id[%i] ver[%i] "
	 *			     "frag[%i]\n",
	 *			     inet_ntoa(*((struct in_addr*) &dst_ip )),
	 *				tos, id, ver, frag_off);
	 *		}
	 */
		}

		if ( !(acx % 1000) ) {
			if ( acx == 0 )
				continue;
			gettimeofday(&tv2, NULL);
			sec = (tv2.tv_sec - tv.tv_sec)
			      - (tv.tv_usec - tv2.tv_usec) / 1000000.0;
			printf(" %li @ %.1f pkts/sec and %.1f k/s\n", acx,
				1000/sec, (datapushed / 1024.0) / sec);
			datapushed=0;
			gettimeofday(&tv, NULL);
		}


		/* Flood protection for low traffic limit only. */
		if ( max_pushed < 10000000 ) {
			gettimeofday(&tv2, NULL);
			sec = (tv2.tv_sec - tv.tv_sec)
		      		- (tv.tv_usec - tv2.tv_usec) / 1000000.0;
			if ( (datapushed / sec) >= max_pushed )
				usleep(10);	/* 10 should give up our timeslice */
		}
	}


	gettimeofday(&tv, NULL);
	printf("\nWrote %li packets in %.2fs @ %.2f pkts/s\n", acx,
		(tv.tv_sec-starttime.tv_sec)
		+ (tv.tv_usec-starttime.tv_usec) / 1000000.0,
		acx / ((tv.tv_sec-starttime.tv_sec)
                       + (tv.tv_usec-starttime.tv_usec)/1000000.0) );

	libnet_destroy(l);
	free(buf);
	return ( 0 );
}

void usage(char *name)
{
   fprintf(stderr,
	"usage: %s [-v] [-D] -s <source ip> -d <destination ip> [-r <random seed>]\n"
	"          [-p <pkts to generate>] [-k <skip packets>] [-x <repeat times>]\n"
	"          [-m <max kB/s to generate>]\n\n"
	"	Percentage Opts: [-F frags] [-V <Bad IP Version>]\n"
	"		         [-I <Random IP Header length>]\n"
	"notes:\n"
	"	[-D] causes packet info to be printed out -- DEBUGGING\n\n"
	"       ex: -s a.b.c.d   -d a.b.c.d -F100\n"
	"        100%% of the packets will be ^^^^ fragments\n"
	"       ex: -s a.b.c.d   -d a.b.c.d -p 100 -r 103334\n"
	"       ex: -s rand -d rand -r 23342\n"
	"              ^^^^ causes random source addr\n"
	"       ex: -s rand -d rand -k 10000 -p 10001 -r 666\n"
	"           Will only send the 10001 packet with random seed 666\n"
	"           this is especially useful if you suspect that packet is\n"
	"           causing a problem with the target stack.\n"
	"\n",
	((char *) rindex(name, '/')) == ((char *) NULL)
		? (char *) name
		: (char *) rindex(name, '/') + 1);
}

void sighandler(int sig)
{
	struct timeval tv;
	gettimeofday(&tv, NULL);

	printf("\n");
	printf("Caught signal %i\n", sig);

	printf("Used random seed %i\n", seed);
	printf("Wrote %li packets in %.2fs @ %.2f pkts/s\n", acx,
		(tv.tv_sec - starttime.tv_sec)
		  + (tv.tv_usec - starttime.tv_usec)/1000000.0,
		acx / (( tv.tv_sec - starttime.tv_sec)
		  + (tv.tv_usec - starttime.tv_usec)/1000000.0)
		);

	fflush(stdout);
	exit(0);
}
